File: Max-Area-of-Island.py
1import queue
2class Solution:
3    def bfs(self, grid: List[List[int]], row: int, col: int,R: int,C: int) -> int:
4        diff = [0,1,0,-1,0]
5        grid[row][col] = 0
6        q = queue.Queue()
7        q.put([row,col])
8
9        area = 1
10        while(not q.empty()):
11            val = q.get()
12            currentrow, currentcol = val[0],val[1]
13            for di in range(4):
14                adjR,adjC = currentrow + diff[di], currentcol + diff[di+1]
15                if(adjR>=0 and adjR<R and adjC>=0 and adjC<C and grid[adjR][adjC] == 1):
16                    q.put([adjR,adjC])
17                    grid[adjR][adjC] = 0
18                    area+=1
19        return area
20
21    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
22        R,C = len(grid), len(grid[0])
23        currarea,maxarea = 0,0
24        for row in range(R):
25            for col in range(C):
26                if(grid[row][col] == 1):
27                    currarea = self.bfs(grid,row,col,R,C)
28                    maxarea = max(maxarea,currarea)
29        return maxarea
30

File: Max-Area-of-Island.java
1class Solution {
2    static int[] diff = {0,1,0,-1,0};
3    private int bfs(int[][] grid, int row, int col, int R, int C){
4        grid[row][col] = 0;
5        Queue<int[]> queue = new LinkedList<>();
6        queue.add(new int[]{row,col});
7        int area = 1;
8        while(!queue.isEmpty()){
9            int[] val = queue.poll();
10            int currentrow = val[0] , currentcol = val[1];
11            for(int di = 0 ; di < 4 ; di++){
12                int adjR = currentrow + diff[di], adjC = currentcol + diff[di+1];
13                if(adjR>=0 && adjR<R && adjC>=0 && adjC<C && grid[adjR][adjC] == 1){
14                    grid[adjR][adjC] = 0;
15                    queue.add(new int[]{adjR,adjC});
16                    area++;
17                }
18            }
19        }
20        return area;
21    }
22    public int maxAreaOfIsland(int[][] grid) {
23        int R = grid.length, C = grid[0].length;
24        int currarea = 0, maxarea = 0;
25        for(int row = 0 ; row < R ; row++){
26            for(int col = 0 ; col < C ; col++){
27                if(grid[row][col] == 1){
28                    currarea = bfs(grid,row,col,R,C);
29                    maxarea = Math.max(maxarea,currarea);
30                }
31            }
32        }
33        return maxarea;
34    }
35}